extensions/gitpod-remote/src/machineSettings.ts

```diff
@@ -0,0 +1,59 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { parse as parseJson, modify as modifyJson, applyEdits as applyEditsJson } from 'jsonc-parser';
+import { GitpodExtensionContext } from 'gitpod-shared';
+import { PortOnOpen, WorkspaceConfig } from '@gitpod/gitpod-protocol/lib/protocol';
+
+export async function configureMachineSettings(context: GitpodExtensionContext, wsConfig: WorkspaceConfig) {
+	const extRemoteLogsUri: vscode.Uri = context.logUri;
+	const remoteUserDataPath = path.posix.dirname(path.posix.dirname(path.posix.dirname(path.posix.dirname(context.logUri.path))));
+	const machineSettingsResource = extRemoteLogsUri.with({ path: path.posix.join(remoteUserDataPath, 'Machine', 'settings.json') });
+	try {
+		let settingsStr: string = '{}';
+		const fileExists = await vscode.workspace.fs.stat(machineSettingsResource).then(() => true, () => false);
+		if (fileExists) {
+			const settingsbuffer = await vscode.workspace.fs.readFile(machineSettingsResource);
+			settingsStr = new TextDecoder().decode(settingsbuffer);
+		}
+
+		// settings.json is a json with comments file so we use jsonc-parser library
+		let modified = false;
+		const settingsJson = parseJson(settingsStr);
+		if (settingsJson['remote.autoForwardPortsSource'] === undefined) {
+			const edits = modifyJson(settingsStr, ['remote.autoForwardPortsSource'], 'process', { formattingOptions: { insertSpaces: true, tabSize: 4, insertFinalNewline: true } });
+			settingsStr = applyEditsJson(settingsStr, edits);
+			modified = true;
+		}
+		if (settingsJson['remote.portsAttributes'] === undefined && wsConfig.ports?.length) {
+			const mapOnOpen = (onOpen?: PortOnOpen) => {
+				switch (onOpen) {
+					case 'open-browser': return 'openBrowser';
+					case 'open-preview': return 'openPreview';
+					default: return onOpen;
+				}
+			};
+
+			const portsAttributes: any = {};
+			for (const portConfig of wsConfig.ports) {
```

Comment:
it should be based on info from https://github.com/gitpod-io/gitpod/blob/39c5542024f7717041efde80cd1d5476e725ca7b/components/supervisor-api/status.proto#L145

it can change dynamically as a user edits .gitpod.yml